File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> answer;
        backtracking(answer, "", 0, 0, n);
        return answer;
    }

    void backtracking(vector<string>& answer, string curString, int leftCount, int rightCount, int n) {
        if (curString.size() == 2 * n) {
            answer.push_back(curString);
            return;
        }
        if (leftCount < n) {
            curString += '(';
            backtracking(answer, curString, leftCount + 1, rightCount, n);
            curString.pop_back();
        }
        if (leftCount > rightCount) {
            curString += ')';
            backtracking(answer, curString, leftCount, rightCount + 1, n);
            curString.pop_back();
        }
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        l = 0
        r = 0

        res = []
        
        def dfs(l,r,part):
            if l == n and r == n:
                res.append(part[:])
                return

            if r+1<=n and l > r:
                dfs(l,r+1,part+")")

            if l+1<=n:
                dfs(l+1,r,part+"(")
        
        dfs(0,0,"")
        return res
            